### ISSUE TITLE
add output of `pip freeze` to verdi export?

### ISSUE THREAD

**Issue by ltalirz** [aiidateam/aiida-core#3363]
In order to reproduce things in the python ecosystem, knowing the AiiDA version ([which is what people currently report on the archive](https://archive.materialscloud.org/2018.0001/v3)) isn't quite enough - in the current way of packaging, even packages like AiiDA that are very strict on dependencies can break when dependencies of dependencies change (we've had examples of this in the past).

In order to remedy this, it might be helpful to keep the output of `pip freeze` around; and one good place to put it could be AiiDA export files.
As a first step, we could simply put a `requirements.txt` in the export file for reference. If people start using this, we could add more tools around it later.

Mentioning @giovannipizzi @CasperWA for comment

P.S. This came out of a discussion with @greschd  in [another thread](https://github.com/greschd/aiida-strain/issues/1#issuecomment-535709684)

**Comment by sphuber** [comment]
If this is to provide information as to with which packages (and versions thereof) the exported nodes have been generated, I don't think this is a good solution. One might generate some nodes with a certain package and then update it later on. Trying to `pip freeze` output at the time of export for this reason most likely will be more confusing than it will help. Note that we already recently added the functionality that now automatically will add the version of `aiida-core` and that of the plugin (if applicable to the process class) to process nodes, to register with what version they were generated.

**Comment by ltalirz** [comment]
> One might generate some nodes with a certain package and then update it later on.

This is correct, of course - I still think it is a step in the right direction to keep the versions of packages at export.

> Note that we already recently added the functionality that now automatically will add the version of aiida-core and that of the plugin (if applicable to the process class) to process nodes, to register with what version they were generated.

That is good but, unfortunately, not enough.
It is very likely that updates of dependencies (or dependencies of dependencies) break e.g. aiida-core installations - try installing any of the old 0.x versions for example.
Note that this also applies to basically all aiida plugins that were released before aiida-core 1.0, since they depended on AiiDA using something like `aiida-core>=0.x`.

I think it would be useful for a user to have a central place where he or she can see which versions of packages need to be installed.


 Trying to pip freeze output at the time of export for this reason most likely will be more confusing than it will help.

**Comment by sphuber** [comment]
> Trying to pip freeze output at the time of export for this reason most likely will be more confusing than it will help.

Wait, I am confused. Are you now saying you think it _will_ or _won't_ be useful to include `pip freeze` output in export archives?

**Comment by ltalirz** [comment]
sorry, I copied this from your post and forgot to delete it ;-)

**Comment by sphuber** [comment]
Ah ok fair enough. But I still am not sure if it really will be useful. An AiiDA database will almost certainly contain nodes that have been created with many possible environmental configurations. Including the particular state of the python packages at the time of exporting as a way of recording what was applicable for the exported nodes seems pointless as it will potentially not apply to even a single node in the export. They all may have been generated with a different python environment. So the chances are that recreating the environment from that `pip freeze` will not allow you to reproduce those nodes whatsoever.

**Comment by ltalirz** [comment]
Well, I guess the key question here is: will it be useful in practice? (even though there are no hard guarantees)
I would say yes.

**Comment by giovannipizzi** [comment]
As it doesn't harm, maybe we can include it in the export file (and document it in the correct page, maybe mentioning it's a feature that is there just for "logging" purposes and might change in the future). I don't think we need to provide functionality for now on how to get those (unless we agree) - we just say in the docs that if you really need that piece of info, you can uncompress the zip .aiida file and get it. What do you think?

**Comment by sphuber** [comment]
Adding some minimum tooling would not be too difficult. I already built the `verdi export inspect` endpoint sometime ago that can be used to inspect the contents. We could easily add a `-r` flag to echo the contents of the `requirements.txt` file if present.

I just don't think that the "does no harm" is necessarily true. Putting this file there seems to suggest that it can be useful, even if we very explicitly say there are no guarantees. Is this primarily aimed to know what the dependencies of our dependencies were? Or more of our own plugins? What is the point in knowing that a file was exported with `PyCIFRW==4.2.1` if very possibly, none of the nodes in the export file were created using that version?

**Comment by giovannipizzi** [comment]
> What is the point in knowing that a file was exported with PyCIFRW==4.2.1 if very possibly, none of the nodes in the export file were created using that version?

That's a good point (and could leak info on what the user has installed on the venv).
On the other hand, it says that in that venv at least things were working (so not an information on who created it - for this there is the version in the attributes that you implemented - but on one of the possible environments that work if you want to reproduce it).

Anyway, I'm ok with either choice (putting or not putting it), still with a little bias on putting it.

**Comment by ltalirz** [comment]
>  it says that in that venv at least things were working

I think this is the main point here - to provide a set of packages that is working together and can be used to continue working with the data set.

If we were continuously testing to pip install all previously released versions of AiiDA (& plugins), then this would be unnecessary, but currently we don' t do that.

**Comment by sphuber** [comment]
> I think this is the main point here - to provide a set of packages that is working together and can be used to continue working with the data set.

This is not true though. The export can contain nodes that were generated with older versions of certain packages and trying to access them with the current venv might very well be broken. The fact that you can export certain nodes with the current venv does not mean that their database state is compatible with the code. It might be that some of those nodes have never been loaded and tested with the current environment.
